### ISSUE TITLE
PMIX_RANK_UNDEF doesn't work with PMIx_Get()

### ISSUE THREAD

**Issue by RaymondMichael** [openpmix/openpmix#3259]
### What version of the PMIx Reference Library are you using? 

4.2.6

### Describe how PMIx was installed 

Running with HPE's PALS launcher.

### Please describe the system on which you are running

Running with 1 host and 2 ranks.

## Details of the problem

In the following program, the PMIx_Get() fails to find the key when the passed proc.rank is PMIX_RANK_UNDEF. When proc.rank is changed to the rank that posted the key, it finds it fine. According to the spec, this should work.

#include <assert.h>
#include <stdio.h>
#include <pmix.h>
#include <string.h>

int
main(int argc, char *argv[])
{
    pmix_value_t *val = NULL;
    pmix_status_t rc = PMIX_SUCCESS;
    pmix_proc_t global_proc, proc;

    rc = PMIx_Init(&global_proc, NULL, 0);
    assert(PMIX_SUCCESS == rc);

    rc = PMIx_Get(&global_proc, PMIX_LOCAL_RANK, NULL, 0, &val);
    assert(PMIX_SUCCESS == rc);

    int local_rank = val->data.uint16;
    PMIX_VALUE_RELEASE(val);

    char key_p[32];
    pmix_value_t value_p;

    sprintf(key_p, "%s-%d", "foo", local_rank);
    value_p.type = PMIX_UINT32;
    value_p.data.uint32 = local_rank + 10;

    rc = PMIx_Put(PMIX_GLOBAL, key_p, &value_p);
    assert(PMIX_SUCCESS == rc);

    rc = PMIx_Commit();
    assert(PMIX_SUCCESS == rc);

    PMIX_PROC_CONSTRUCT(&proc);
    proc.rank = PMIX_RANK_UNDEF;
    strcpy(proc.nspace, global_proc.nspace);

    for (int i = 0; i < 2; i++) {
        char key_g[32];
        pmix_value_t *value_g;
#if 0
        proc.rank = i;
#endif
        sprintf(key_g, "%s-%d", "foo", i);
        rc = PMIx_Get(&proc, key_g, NULL, 0, &value_g);
        if (PMIX_SUCCESS != rc) {
            fprintf(stderr, "PMIx_Get got %d\n", rc);
        } else {
            fprintf(stderr, "Got %d\n", value_g->data.uint32);
        }
    }

    PMIx_Finalize(NULL, 0);

    return 0;
}

**Comment by rhc54** [comment]
Hmmm...it works with current head of master branch, and it works if I force use of the `hash` component - so the problem lies in the `dstore` (yet again). Fix should be relatively easy - will try to get it today.

Thx for the heads-up!

**Comment by RaymondMichael** [comment]
I build 4.2.9rc1 this morning and I'm still seeing the problem. When the clients do the PMIx_Get()s, I see the PMIx server print the following. It was doing this with the old code too.

[raymond-rm-1:32120] PMIX ERROR: PMIX_ERR_NOT_FOUND in file dstore_base.c at line 1567
[raymond-rm-1:32120] PMIX ERROR: PMIX_ERROR in file dstore_base.c at line 2334
[raymond-rm-1:32120] PMIX ERROR: PMIX_ERR_NOT_FOUND in file dstore_base.c at line 1567
[raymond-rm-1:32120] PMIX ERROR: PMIX_ERROR in file dstore_base.c at line 2334

**Comment by rhc54** [comment]
Huh - okay, let me take another look. Sorry for the problem.

**Comment by rhc54** [comment]
Please give it another try - hopefully got the last spot now.

**Comment by RaymondMichael** [comment]
Same problem and messages. Is there logging I can turn on to help you?

**Comment by rhc54** [comment]
?? How odd - it worked for me using your test code on Linux. You pulled the head of the v4.2 branch? Should be at hash 0x604566a

**Comment by RaymondMichael** [comment]
Yes. I patched our internal clone with everything in the v4.2 branch and replaced the RPMs on our machine with freshly built ones. Is there something I can do to make certain I have your code?

**Comment by rhc54** [comment]
The change was in the `src/server/pmix_server.c` code - you should see this code block start at line 4514:
```c
         /* if the peer is using the "dstore" component, then
         * we also have to send back any session/node/app-level
         * info so it can be stored locally in their hash */
        if (0 != strcmp("hash", peer->nptr->compat.gds->name)) {
            PMIX_GDS_FETCH_INFO_ARRAYS(rc, pmix_globals.mypeer, reply);
            if (PMIX_SUCCESS != rc) {
                PMIX_ERROR_LOG(rc);
                PMIX_RELEASE(reply);
                return rc;
            }
            // and we have to ensure their hash component knows
            // the job size
            PMIX_CONSTRUCT(&cb, pmix_cb_t);
            PMIX_LOAD_PROCID(&proc, peer->info->pname.nspace, PMIX_RANK_WILDCARD);
            cb.proc = &proc;
            cb.scope = PMIX_INTERNAL;
            cb.copy = false;
            cb.key = PMIX_JOB_SIZE;
            PMIX_GDS_FETCH_KV(rc, pmix_globals.mypeer, &cb);
            if (PMIX_SUCCESS == rc) {
                kv = (pmix_kval_t*)pmix_list_remove_first(&cb.kvs);
                PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, reply, kv, 1, PMIX_KVAL);
                PMIX_RELEASE(kv);
            } else {
                // try to retrieve from another spot
                PMIX_GDS_FETCH_KV(rc, peer, &cb);
                if (PMIX_SUCCESS == rc) {
                    kv = (pmix_kval_t*)pmix_list_remove_first(&cb.kvs);
                    PMIX_BFROPS_PACK(rc, pmix_globals.mypeer, reply, kv, 1, PMIX_KVAL);
                    PMIX_RELEASE(kv);
                }
            }
            PMIX_DESTRUCT(&cb);
        }
 ```

**Comment by RaymondMichael** [comment]
I've got that in what I'm building. Just in case, we're doing this with both ranks on the same host. PALS doesn't support the dmodex handler yet, but that shouldn't matter since we're not using multiple hosts.

**Comment by rhc54** [comment]
Yeah, I'm running on the same host too. Let me try again and see what might be happening. Just to be sure: PALS is providing the `PMIX_JOB_SIZE` in the register_nspace call, yes? It can be in a job-info array or standalone, but needs to be there.

**Comment by RaymondMichael** [comment]
Yes. PMIX_JOB_SIZE is going into the top level info for _register_nspace() and I verified it's getting set to 2.

FWIW my PMIx_Get() calls for the other rank's key is returning -24 "timed out". I can understand that one of the two might get this, but at least one of them should always get both keys.

**Comment by RaymondMichael** [comment]
I just copied the example program to make sure, and it's still getting -24. I agree that from the spec, that shouldn't be happening.

**Comment by rhc54** [comment]
Very strange - I rebuilt from scratch and it still works correctly for me. If you go to the `test/simple` directory under PMIx, you should see a `legacy.c` file - this is just your test code. You'll also see `simptest.c` - this is a real simple server. If you make those to get the binaries, and execute `./simptest -n 2 -e ./legacy`, does that work?

**Comment by RaymondMichael** [comment]
That seemed to work.

Testing version OpenPMIx 4.2.9a1 (PMIx Standard: 4.2, Stable ABI: 0.0, Provisional ABI: 0.0)
Collecting inventory
Inventory collected: 0
Got 10
Got 11
PMIx_Get got -24
Got 11
SIMPTEST: Model event handler called with status -147(PMIX_MODEL_DECLARED)
	pmix.pgm.model:	PMIX
	pmix.mdl.name:	test
	pmix.evname:	SIMPTEST-MODEL
Test finished OK!

**Comment by rhc54** [comment]
Hmmm...not entirely - you still got that odd timeout response for one of the "get" calls. Here is what you should see:
```shell
$ ./simptest -n 2 -e ./legacy
Testing version OpenPMIx 4.2.9a1, repo rev: v4.2.8-6-g604566a (PMIx Standard: 4.2, Stable ABI: 0.0, Provisional ABI: 0.0)
Collecting inventory
Inventory collected: 0
Got 10
Got 10
Got 11
Got 11
SIMPTEST: Model event handler called with status -147(PMIX_MODEL_DECLARED)
	pmix.pgm.model:	PMIX
	pmix.mdl.name:	test
	pmix.evname:	SIMPTEST-MODEL
Test finished OK!
$
```
I wonder what is causing the difference? I'm running on Centos 7 with an old gcc 4.8.5. I imagine you are running on something more modern?

**Comment by RaymondMichael** [comment]
I'm on sles15sp5. Just to make sure it wasn't something in our own branch, I grabbed a pure copy of your branch and had the same thing. Maybe I should start adding some printfs to trace what's going on.

**Comment by rhc54** [comment]
You can put `PMIX_MCA_pmix_server_get_verbose=5` in the environment to see what the server-side is doing. You can do the same for the client side, replacing "server" with "client".

**Comment by RaymondMichael** [comment]
[log.txt](https://github.com/openpmix/openpmix/files/14056276/log.txt)
Here you go.

**Comment by rhc54** [comment]
Very interesting - I see that one proc is running quite a bit ahead of the other. The time difference is causing the internal timeout registration. Let me force that issue locally and see if I can reproduce.

**Comment by rhc54** [comment]
One thing you might try: add a `PMIx_Fence` after the `PMIx_Commit` in your test code. I suspect that will fix the problem. I guess we technically don't require that when on a single node (would have to check), and I'll take a look here anyway. Just would confirm that the issue is the race condition.

**Comment by RaymondMichael** [comment]
I did a non-PMIX_COLLECT_DATA fence and it made things work. According to the standard, I don't think it should be needed on multi-host either.

**Comment by rhc54** [comment]
Yeah, I agree - though it will lead to dmodex calls and probable timeouts at scale. Still, it confirms the local race condition, so I'll try to track that down.

Thx!

**Comment by RaymondMichael** [comment]
A certain customer, whom you're well aware off, wants to use this as a global key-value store. They don't know which rank will be posting the global key, and so they're using PMIX_RANK_UNDEF. They know about PMIx_Fence(), but at scale it's going to be expensive, and so they were hoping for something better. Since they have a good guess of the subset of processes that will publish the key, I suggested they look at process groups to create a new namespace for a smaller fence.

One alternative is to use the pub / lookup handlers to connect to an actual centralized key-value store.

Either way, it sounds like an amendment to the standard might be needed to keep people away from this.

**Comment by rhc54** [comment]
Ah, indeed. You're suggestion of process groups would help, assuming your backend knows how to efficiently execute a fence across the subset. People have asked about using pub/lookup more generally, but tests show it performs poorly when used at scale - of course, that's with PRRTE as the backend. You might have a more efficient backend. Still, it involves individual messages being sent to the pub/lookup server, so it isn't a generally great solution.

FWIW: I had at one time explored implementing the pub/lookup server as a distributed hash table to try and both improve performance and deal with fault tolerance. It actually isn't bad, though it would take a little more effort to bring it to completion. I suppose one could put the DHT portion of the code in PMIx (under the pub/lookup interfaces) and have the host send the inter-node requests - or just leave it all to the hosts to implement. Dunno - might be worth pondering if you think this will be a consistent request.

I'll suggest the Standards folks add something to make this clearer, and will see if I can perhaps improve that timeout support. At least I know it is working! May want it to be a little less restrictive.

**Comment by RaymondMichael** [comment]
A sentence in section 5.1 may handle this. "In addition, the keys must be globally exchanged prior to retrieval as there is no way for the host to otherwise locate the source for the information."

**Comment by rhc54** [comment]
👍

**Comment by rhc54** [comment]
Found there were some bugs in how we resolved cached data requests at the server, so I cleaned that up. Your race condition is now resolved. Still won't help with the global scaling issue or the need for the fence, but at least this now works cleanly.

**Comment by RaymondMichael** [comment]
The team is curious what "direct modex" means.
In my own testing I've modified the test program to have each rank use a different key. Having the processes update the same key was producing some randomness.

**Comment by rhc54** [comment]
> Having the processes update the same key was producing some randomness.

Absolutely - the key must be globally unique, meaning that nobody else can "put" it.  Otherwise, you have no way of knowing which version of the key to return since you aren't referencing it by its publisher.

> what "direct modex" means.

Direct modex means that you request the value of the key from the daemon hosting the proc that published it. Instead of doing a fence and collectively scooping all the keys up, you do a remote lookup on a per-request basis. So when a local proc requests a key from a specific target proc, the server asks its host to contact whatever daemon is hosting that target and return the value for that key.

It is expected that the remote daemon will wait until the target publishes the requested key before responding, so we advise people provide a `PMIX_TIMEOUT` if they suspect a direct modex is going to be involved. You usually know that because you skip the fence.

Doing all those direct requests hurts scalability, of course - much more efficient to gather everything in a collective. One optimization that can help is for the remote daemon to return all data published by the target - we assume that one request for a key will probably be followed by a request for a different key. You can also have the remote daemon return all data published by all local procs from that nspace - again, you assume that if someone from the original node asked about one of the local procs, someone else from that node will likely ask about one or more of the local procs as well.

Depending on the pattern, you can gain some improvements with those steps. Still not as good as a collective, but it can help a lot. PRRTE tries to play that game to an extent.

**Comment by RaymondMichael** [comment]
Useful information, but we were actually wondering about the word "modex". I can't recall having heard the word anywhere else. :)

**Comment by rhc54** [comment]
Ah, indeed. So Argonne calls the exchange of data the "business card exchange". The OMPI community calls it the "module exchange", shortened to "modex", because the sources of the information being exchanged are the individual "modules" (a.k.a plugins) within the OMPI code base. We've been calling it that for over 20 years now, so it is just built into our lexicon. 

Since PMIx (and PRRTE) is built on the same plugin-based architecture, it just was natural to use the term. 🤷‍♂️
